cpp/locality/NeighborComputeFunctional.cc

```diff
@@ -15,8 +15,14 @@ NeighborList makeDefaultNlist(const NeighborQuery* nq, const NeighborList* nlist
         auto nqiter(nq->query(query_points, num_query_points, qargs));
         nlist = nqiter->toNeighborList();
     }
+    // Ideally we wouldn't allocate a new NeighborList at all, but we don't want to force
+    // calling code to manage the memory of a raw pointer, so we prefer to return by value
+    // here, and the toNeighborList function has to return a raw pointer to work with the
+    // Cython interface of freud. Therefore, for now we're sticking with this suboptimal
+    // solution until we can refactor these internalsa little more cleanly.
```

Comment:
```suggestion
    // solution until we can refactor these internals a little more cleanly.
```

Reply:
@vyasr What is the "until we can refactor" here? Is it a matter of developer time/effort or some other technical constraint? What would need to be done?

Reply:
The technical constraint is that `toNeighborList` needs to return a raw pointer for Cython, while `makeDefaultNlist` promises a `NeighborList`,  not a pointer. The latter choice simplifies code in all the compute classes that use this function since they can just have a NeighborList object and not think about deleting this pointer, otherwise they'd have to delete it. We can't use a shared pointer directly without making a copy in the case where the user provides a `NeighborList`, which is counterintuitive since that's supposed to be the efficient case.

Reply:
You would need to define a separate function for it. Changing the return value alone is not sufficient for the compiler to differentiate two function signatures.

You could reasonably write an internal function (say `populateNeighborList`) that contains almost all the logic of `toNeighborList` that accepts a pointer to a `NeighborList` and simply writes the data into that `NeighborList`. The existing `toNeighborList` function would make a `new NeighborList()` and pass that directly through to `populateNeighborList`. You could then write a new function that creates a `nl = NeighborList()` on the stack, then calls `populateNeighborList(*nl)` to update this object directly, then return it by value. However, in this case the new function would still be returning by value, so you'd have to look at the precise rules for copy elision in C++ to verify that you can depend on compilers to optimize out the copy operation; otherwise you'd be back where you started, except that the copy operation would be implicit on the function return rather than explicit inside `makeDefaultNlist`.